Discord Messages over 2000 Charaters Fail

The discord client kicks out an error for lengthy responses to things like "write me a contract". These types of responses should provide a link with the full output, summarize the response in a shorter total length, or break the response up into multiple Discord messages.

Example: https://stackoverflow.com/questions/60686321/discord-js-content-must-be-2000-or-fewer-in-length

```2023-12-19 02:10:00,932 [ERROR] discord.client: Ignoring exception in on_message
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/discord/client.py", line 441, in _run_event
    await coro(*args, **kwargs)
  File "/root/DiscordianAI/bot.py", line 312, in on_message
    await message.channel.send(response)
  File "/usr/local/lib/python3.10/dist-packages/discord/abc.py", line 1561, in send
    data = await state.http.send_message(channel.id, params=params)
  File "/usr/local/lib/python3.10/dist-packages/discord/http.py", line 745, in request
    raise HTTPException(response, data)
discord.errors.HTTPException: 400 Bad Request (error code: 50035): Invalid Form Body
In content: Must be 2000 or fewer in length.```

--- a/.flake8
+++ b/.flake8
@@ -1,2 +1,2 @@
 [flake8]
-max-line-length = 79
+max-line-length = 99

--- a/README.md
+++ b/README.md
@@ -2,7 +2,7 @@
 
 # Description
 
-This is a Python script for a Discord bot that uses OpenAI's GPT API to generate responses to user messages. The bot can be configured to listen to specific channels and respond to direct messages. The bot also has a rate limit to prevent spamming and can maintain a per user conversational history to improve response quality which is only limited by the `GPT_TOKENS` value.
+This is a Python script for a Discord bot that uses either OpenAI's GPT API, or any compatible API such as Perplexity to generate responses to user messages. The bot can be configured to listen to specific channels and respond to direct messages. The bot also has a rate limit to prevent spamming and can maintain a per user conversational history to improve response quality which is only limited by the `GPT_TOKENS` value.
 
 # Requirements
 
@@ -30,12 +30,14 @@ The `config.ini` file contains the following configuration sections:
 - `ACTIVITY_TYPE`: The type of activity for the bot (e.g. playing, streaming, listening, watching, custom, competing).
 - `ACTIVITY_STATUS`: The activity status of the bot (e.g. Humans).
 
-### OpenAI
+### Default Configs
 
-- `OPENAI_API_KEY`: The OpenAI API key.
-- `OPENAI_TIMEOUT`: The OpenAI API timeout in seconds. (default: `30`)
-- `GPT_MODEL`: The GPT model to use (default: `gpt-3.5-turbo`).
-- `GPT_TOKENS`: The maximum number of tokens to generate in the GPT response (default: `3072`).
+- `API_URL`: The backend API URL. (default: `https://api.openai.com/v1/`)
+- `API_KEY`: The API key for your backend. (default: `None`)
+- `GPT_MODEL`: The GPT model to use (default: `gpt-4o-mini`)
+- `INPUT_TOKENS`: Your response input size. (default: `120000`)
+- `OUTPUT_TOKENS`: The maximum number of tokens to generate in the GPT response (default: `8000`)
+- `CONTEXT_WINDOW`: The maximum number of tokens to keep in the context window. (default: `128000`)
 - `SYSTEM_MESSAGE`: The message to send to the GPT model before the user's message.
 
 ### Limits
@@ -59,11 +61,13 @@ BOT_PRESENCE = online
 ACTIVITY_TYPE=listening
 ACTIVITY_STATUS=Humans
 
-[OpenAI]
-OPENAI_API_KEY = <your_openai_api_key>
-OPENAI_TIMEOUT=30
-GPT_MODEL=gpt-3.5-turbo
-GPT_TOKENS=3072
+[Default]
+API_URL=https://api.openai.com/v1/
+API_KEY = <your_api_key>
+GPT_MODEL=gpt-4o-mini
+INPUT_TOKENS=120000
+OUTPUT_TOKENS=8000
+CONTEXT_WINDOW=128000
 SYSTEM_MESSAGE = You are a helpful AI assistant.
 
 [Limits]
@@ -116,6 +120,14 @@ This will execute forever, unless manually stopped. The `-v` option is used to m
 - Customizable configurations: The script allows for different modes of operation and configurations.
 - Error Handling: It logs errors and exits on failure.
 - Process Handling: It terminates existing instances of the bot before starting a new one.
+- Rate Limiting: Implements rate limiting to prevent users from spamming commands.
+- Conversation History: Maintains conversation history for each user to provide context-aware responses.
+- Activity Status: Configurable activity status to display what the bot is doing.
+- Direct Message Handling: Processes direct messages separately from channel messages.
+- Channel Message Handling: Processes messages in specific channels where the bot is mentioned.
+- Automatic Message Splitting: Automatically splits long messages to fit within Discord's message length limits.
+- Global Exception Handling: Catches and logs unhandled exceptions to prevent crashes.
+- Shard Support: Supports sharding for better scalability and performance.
 
 ### Usage
 

--- a/bot.py
+++ b/bot.py
@@ -14,12 +14,20 @@ import discord
 from openai import OpenAI
 from websockets.exceptions import ConnectionClosed
 
+
+# Define the RateLimiter class
 class RateLimiter:
     def __init__(self):
         self.last_command_timestamps = {}
         self.last_command_count = {}
 
-    def check_rate_limit(self, user_id: int, rate_limit: int, rate_limit_per: int, logger: logging.Logger) -> bool:
+    def check_rate_limit(
+        self,
+        user_id: int,
+        rate_limit: int,
+        rate_limit_per: int,
+        logger: logging.Logger
+    ) -> bool:
         current_time = time.time()
         last_command_timestamp = self.last_command_timestamps.get(user_id, 0)
         last_command_count_user = self.last_command_count.get(user_id, 0)
@@ -38,6 +46,7 @@ class RateLimiter:
         logger.info(f"Rate limit exceeded for user: {user_id}")
         return False
 
+
 # Define the function to parse command-line arguments
 def parse_arguments() -> argparse.Namespace:
     try:
@@ -48,6 +57,7 @@ def parse_arguments() -> argparse.Namespace:
     except Exception as e:
         logger.error(f"Error parsing arguments: {e}")
         raise
+
 
 # Define the function to load the configuration
 def load_configuration(config_file: str) -> configparser.ConfigParser:
@@ -67,6 +77,7 @@ def load_configuration(config_file: str) -> configparser.ConfigParser:
     except Exception as e:
         logger.error(f"Error loading configuration: {e}")
         raise
+
 
 def set_activity_status(
     activity_type: str,
@@ -94,6 +105,7 @@ def set_activity_status(
         logger.error(f"Error setting activity status: {e}")
         raise
 
+
 # Define the function to get the conversation summary
 def get_conversation_summary(conversation: list[dict]) -> list[dict]:
     """
@@ -120,6 +132,7 @@ def get_conversation_summary(conversation: list[dict]) -> list[dict]:
         logger.error(f"Error getting conversation summary: {e}")
         raise
 
+
 async def check_rate_limit(
     user: discord.User,
     rate_limiter: RateLimiter,
@@ -139,6 +152,7 @@ async def check_rate_limit(
         logger.error(f"Error checking rate limit: {e}")
         raise
 
+
 async def process_input_message(
     input_message: str,
     user: discord.User,
@@ -152,35 +166,18 @@ async def process_input_message(
 
         INPUT_PROMPT = {"role": "user", "content": input_message}
 
-        conversation = conversation_history.get(user.id, [])
-        conversation.append({"role": "user", "content": input_message})
-
-        conversation_tokens = sum(
-            len(message["content"].split()) for message in conversation
-        )
-
-        if conversation_tokens >= GPT_TOKENS * 0.8:
-            conversation_summary = get_conversation_summary(conversation)
-            conversation_tokens_summary = sum(
-                len(message["content"].split())
-                for message in conversation_summary
-            )
-            max_tokens = GPT_TOKENS - conversation_tokens_summary
-        else:
-            max_tokens = GPT_TOKENS - conversation_tokens
-
-        # Log the current conversation history
-        # logger.info(f"Current conversation history: {conversation}")
+        conversation = CONVERSATION_HISTORY.get(user.id, [])
+        conversation.append(INPUT_PROMPT)
 
         def call_openai_api():
             return client.chat.completions.create(
                 model=GPT_MODEL,
                 messages=[
-                    {"role": "system", "content": SYSTEM_MESSAGE},
+                    SYSTEM_PROMPT,
                     *conversation_summary,
-                    {"role": "user", "content": input_message}
+                    INPUT_PROMPT
                 ],
-                max_tokens=max_tokens,
+                max_tokens=OUTPUT_TOKENS,
                 temperature=0.7
             )
 
@@ -201,14 +198,12 @@ async def process_input_message(
 
         if response_content:
             logger.info("Received response from the API.")
-            # Debugging: Log the raw response
-            # logger.info(f"Raw API response: {response}")
             logger.info(f"Sent the response: {response_content}")
 
             conversation.append(
                 {"role": "assistant", "content": response_content}
             )
-            conversation_history[user.id] = conversation
+            CONVERSATION_HISTORY[user.id] = conversation
 
             return response_content
         else:
@@ -254,13 +249,14 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
         'Default', 'API_URL', fallback='https://api.openai.com/v1/'
     )
     GPT_MODEL = config.get(
-        'Default', 'GPT_MODEL', fallback='gpt-3.5-turbo-1106'
+        'Default', 'GPT_MODEL', fallback='gpt-4o-mini'
     )
-    GPT_TOKENS = config.getint('Default', 'GPT_TOKENS', fallback=4096)
+    INPUT_TOKENS = config.getint('Default', 'INPUT_TOKENS', fallback=120000)
+    OUTPUT_TOKENS = config.getint('Default', 'OUTPUT_TOKENS', fallback=8000)
+    CONTEXT_WINDOW = config.getint('Default', 'CONTEXT_WINDOW', fallback=128000)
     SYSTEM_MESSAGE = config.get(
         'Default', 'SYSTEM_MESSAGE', fallback='You are a helpful assistant.'
     )
-
     RATE_LIMIT = config.getint('Limits', 'RATE_LIMIT', fallback=10)
     RATE_LIMIT_PER = config.getint('Limits', 'RATE_LIMIT_PER', fallback=60)
 
@@ -298,11 +294,11 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     intents.presences = False
 
     # Create a dictionary to store conversation history for each user
-    conversation_history = {}
+    CONVERSATION_HISTORY = {}
 
     # Setup the Bot's Personality
     SYSTEM_PROMPT = {"role": "system", "content": SYSTEM_MESSAGE}
-    
+
     # Create the bot instance
     bot = discord.Client(intents=intents)
 
@@ -387,7 +383,7 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
             return
 
         conversation_summary = get_conversation_summary(
-            conversation_history.get(message.author.id, [])
+            CONVERSATION_HISTORY.get(message.author.id, [])
         )
         response = await process_input_message(
             message.content, message.author, conversation_summary
@@ -414,7 +410,7 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
             return
 
         conversation_summary = get_conversation_summary(
-            conversation_history.get(message.author.id, [])
+            CONVERSATION_HISTORY.get(message.author.id, [])
         )
         response = await process_input_message(
             message.content, message.author, conversation_summary

--- a/bot.py
+++ b/bot.py
@@ -15,9 +15,11 @@ from openai import OpenAI
 from websockets.exceptions import ConnectionClosed
 
 
-# Define the RateLimiter class
 class RateLimiter:
+    """Class to handle rate limiting for users."""
+
     def __init__(self):
+        """Initialize the RateLimiter with empty dictionaries."""
         self.last_command_timestamps = {}
         self.last_command_count = {}
 
@@ -28,6 +30,18 @@ class RateLimiter:
         rate_limit_per: int,
         logger: logging.Logger
     ) -> bool:
+        """
+        Check if a user has exceeded the rate limit.
+
+        Args:
+            user_id (int): The ID of the user.
+            rate_limit (int): The maximum number of allowed commands.
+            rate_limit_per (int): The time period for the rate limit.
+            logger (logging.Logger): The logger instance.
+
+        Returns:
+            bool: True if the user is within the rate limit, False otherwise.
+        """
         current_time = time.time()
         last_command_timestamp = self.last_command_timestamps.get(user_id, 0)
         last_command_count_user = self.last_command_count.get(user_id, 0)
@@ -47,8 +61,13 @@ class RateLimiter:
         return False
 
 
-# Define the function to parse command-line arguments
 def parse_arguments() -> argparse.Namespace:
+    """
+    Parse command-line arguments.
+
+    Returns:
+        argparse.Namespace: Parsed command-line arguments.
+    """
     try:
         parser = argparse.ArgumentParser(description='GPT-based Discord bot.')
         parser.add_argument('--conf', help='Configuration file path')
@@ -59,19 +78,23 @@ def parse_arguments() -> argparse.Namespace:
         raise
 
 
-# Define the function to load the configuration
 def load_configuration(config_file: str) -> configparser.ConfigParser:
+    """
+    Load the configuration from a file or environment variables.
+
+    Args:
+        config_file (str): Path to the configuration file.
+
+    Returns:
+        configparser.ConfigParser: Loaded configuration.
+    """
     try:
         config = configparser.ConfigParser()
 
-        # Check if the configuration file exists
         if os.path.exists(config_file):
             config.read(config_file)
         else:
-            # Fall back to environment variables
-            config.read_dict(
-                {section: dict(os.environ) for section in config.sections()}
-            )
+            config.read_dict({section: dict(os.environ) for section in config.sections()})
 
         return config
     except Exception as e:
@@ -79,12 +102,16 @@ def load_configuration(config_file: str) -> configparser.ConfigParser:
         raise
 
 
-def set_activity_status(
-    activity_type: str,
-    activity_status: str
-) -> discord.Activity:
+def set_activity_status(activity_type: str, activity_status: str) -> discord.Activity:
     """
-    Return discord.Activity object with specified activity type and status
+    Return discord.Activity object with specified activity type and status.
+
+    Args:
+        activity_type (str): The type of activity.
+        activity_status (str): The status of the activity.
+
+    Returns:
+        discord.Activity: The activity object.
     """
     try:
         activity_types = {
@@ -96,9 +123,7 @@ def set_activity_status(
             'competing': discord.ActivityType.competing
         }
         return discord.Activity(
-            type=activity_types.get(
-                activity_type, discord.ActivityType.listening
-            ),
+            type=activity_types.get(activity_type, discord.ActivityType.listening),
             name=activity_status
         )
     except Exception as e:
@@ -106,26 +131,24 @@ def set_activity_status(
         raise
 
 
-# Define the function to get the conversation summary
 def get_conversation_summary(conversation: list[dict]) -> list[dict]:
     """
-    Conversation summary from combining user messages and assistant responses
+    Get a summary of the conversation.
+
+    Args:
+        conversation (list[dict]): The conversation history.
+
+    Returns:
+        list[dict]: The summarized conversation.
     """
     try:
         summary = []
-        user_messages = [
-            message for message in conversation if message["role"] == "user"
-        ]
-        assistant_responses = [
-            message for message in conversation if message["role"] == "assistant"
-        ]
+        user_messages = [msg for msg in conversation if msg["role"] == "user"]
+        assistant_responses = [msg for msg in conversation if msg["role"] == "assistant"]
 
-        # Combine user messages and assistant responses into a summary
-        for user_message, assistant_response in zip(
-            user_messages, assistant_responses
-        ):
-            summary.append(user_message)
-            summary.append(assistant_response)
+        for user_msg, assistant_resp in zip(user_messages, assistant_responses):
+            summary.append(user_msg)
+            summary.append(assistant_resp)
 
         return summary
     except Exception as e:
@@ -142,6 +165,16 @@ async def check_rate_limit(
 ) -> bool:
     """
     Check if a user has exceeded the rate limit for sending messages.
+
+    Args:
+        user (discord.User): The user to check.
+        rate_limiter (RateLimiter): The rate limiter instance.
+        rate_limit (int): The maximum number of allowed commands.
+        rate_limit_per (int): The time period for the rate limit.
+        logger (logging.Logger, optional): The logger instance. Defaults to None.
+
+    Returns:
+        bool: True if the user is within the rate limit, False otherwise.
     """
     if logger is None:
         logger = logging.getLogger(__name__)
@@ -160,6 +193,14 @@ async def process_input_message(
 ) -> str:
     """
     Process an input message using the GPT model.
+
+    Args:
+        input_message (str): The input message from the user.
+        user (discord.User): The user who sent the message.
+        conversation_summary (list[dict]): The conversation summary.
+
+    Returns:
+        str: The response from the GPT model.
     """
     try:
         logger.info("Sending prompt to the API.")
@@ -185,11 +226,8 @@ async def process_input_message(
             )
 
         response = await asyncio.to_thread(call_openai_api)
-        # Debugging: Log the raw response
-        # logger.info(f"Raw API response: {response}")
 
         try:
-            # Extracting the response content from the new API response format
             if response.choices:
                 response_content = response.choices[0].message.content.strip()
             else:
@@ -200,13 +238,9 @@ async def process_input_message(
 
         if response_content:
             logger.info("Received response from the API.")
-            # Debugging: Log the raw response
-            # logger.info(f"Raw API response: {response}")
             logger.info(f"Sent the response: {response_content}")
 
-            conversation.append(
-                {"role": "assistant", "content": response_content}
-            )
+            conversation.append({"role": "assistant", "content": response_content})
             CONVERSATION_HISTORY[user.id] = conversation
 
             return response_content
@@ -228,6 +262,9 @@ async def process_input_message(
 async def process_dm_message(message: discord.Message):
     """
     Process a direct message.
+
+    Args:
+        message (discord.Message): The direct message received.
     """
     logger.info(f'Received DM from {message.author}: {message.content}')
 
@@ -250,6 +287,9 @@ async def process_dm_message(message: discord.Message):
 async def process_channel_message(message: discord.Message):
     """
     Process a message in a channel.
+
+    Args:
+        message (discord.Message): The message received in a channel.
     """
     logger.info(
         'Received message in {} from {}: {}'.format(
@@ -277,26 +317,45 @@ async def process_channel_message(message: discord.Message):
 
 async def send_split_message(channel: discord.abc.Messageable, message: str):
     """
-    Send a message to a channel. If the message is longer than 2000 characters,
-    it is split into multiple messages at the nearest newline character around
-    the middle of the message.
+    Send a message to a channel, splitting it if necessary.
+
+    Args:
+        channel (discord.abc.Messageable): The channel to send the message to.
+        message (str): The message to send.
     """
     if len(message) <= 2000:
         await channel.send(message)
     else:
-        # Find the nearest newline character around the middle of the message
+        # Find the middle index
         middle_index = len(message) // 2
-        split_index = message.rfind('\n', 0, middle_index)
-        if split_index == -1:  # No newline character found
-            split_index = middle_index  # Split at the middle of the message
-        # Split the message into two parts
-        message_part1 = message[:split_index]
-        message_part2 = message[split_index:]
-        # Send the two parts as separate messages
-        await channel.send(message_part1)
-        await channel.send(message_part2)
 
-# Executes the argparse code only when the file is run directly
+        # Find the nearest newline before the middle index
+        split_index = message.rfind('\n', 0, middle_index)
+        if split_index == -1:
+            split_index = middle_index
+
+        # Adjust split index to avoid splitting code blocks
+        before_split = message[:split_index]
+        after_split = message[split_index:]
+
+        # Check if the split occurs within a code block
+        if before_split.count('```') % 2 != 0:
+            # Find the next newline after the middle index
+            split_index = message.find('\n', middle_index)
+            if split_index == -1:
+                split_index = middle_index
+
+            before_split = message[:split_index]
+            after_split = message[split_index:]
+
+        # Ensure no leading/trailing whitespace
+        message_part1 = before_split.strip()
+        message_part2 = after_split.strip()
+
+        await channel.send(message_part1)
+        await send_split_message(channel, message_part2)
+
+
 if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     # Parse command-line arguments
     args = parse_arguments()
@@ -306,36 +365,20 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
 
     # Retrieve configuration details from the configuration file
     DISCORD_TOKEN = config.get('Discord', 'DISCORD_TOKEN')
-    ALLOWED_CHANNELS = config.get(
-        'Discord', 'ALLOWED_CHANNELS', fallback=''
-        ).split(',')
+    ALLOWED_CHANNELS = config.get('Discord', 'ALLOWED_CHANNELS', fallback='').split(',')
     BOT_PRESENCE = config.get('Discord', 'BOT_PRESENCE', fallback='online')
-
-    # ACTIVITY_TYPE playing, streaming, listening, watching, custom, competing
-    ACTIVITY_TYPE = config.get(
-        'Discord', 'ACTIVITY_TYPE', fallback='listening'
-        )
-    ACTIVITY_STATUS = config.get(
-        'Discord', 'ACTIVITY_STATUS', fallback='Humans'
-        )
-
-    # Set your Defaults
+    ACTIVITY_TYPE = config.get('Discord', 'ACTIVITY_TYPE', fallback='listening')
+    ACTIVITY_STATUS = config.get('Discord', 'ACTIVITY_STATUS', fallback='Humans')
     API_KEY = config.get('Default', 'API_KEY')
-    API_URL = config.get(
-        'Default', 'API_URL', fallback='https://api.openai.com/v1/'
-    )
-    GPT_MODEL = config.get(
-        'Default', 'GPT_MODEL', fallback='gpt-4o-mini'
-    )
+    API_URL = config.get('Default', 'API_URL', fallback='https://api.openai.com/v1/')
+    GPT_MODEL = config.get('Default', 'GPT_MODEL', fallback='gpt-4o-mini')
     INPUT_TOKENS = config.getint('Default', 'INPUT_TOKENS', fallback=120000)
     OUTPUT_TOKENS = config.getint('Default', 'OUTPUT_TOKENS', fallback=8000)
     CONTEXT_WINDOW = config.getint('Default', 'CONTEXT_WINDOW', fallback=128000)
     SYSTEM_MESSAGE = config.get(
-        'Default', 'SYSTEM_MESSAGE', fallback='You are a helpful assistant.'
-    )
+        'Default', 'SYSTEM_MESSAGE', fallback='You are a helpful assistant.')
     RATE_LIMIT = config.getint('Limits', 'RATE_LIMIT', fallback=10)
     RATE_LIMIT_PER = config.getint('Limits', 'RATE_LIMIT_PER', fallback=60)
-
     LOG_FILE = config.get('Logging', 'LOG_FILE', fallback='bot.log')
     LOG_LEVEL = config.get('Logging', 'LOG_LEVEL', fallback='INFO')
 
@@ -344,13 +387,9 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     logger.setLevel(getattr(logging, LOG_LEVEL.upper()))
 
     # File handler
-    file_handler = RotatingFileHandler(
-        LOG_FILE, maxBytes=5 * 1024 * 1024, backupCount=5
-    )
+    file_handler = RotatingFileHandler(LOG_FILE, maxBytes=5 * 1024 * 1024, backupCount=5)
     file_handler.setLevel(getattr(logging, LOG_LEVEL.upper()))
-    file_formatter = logging.Formatter(
-        '%(asctime)s [%(levelname)s] %(name)s: %(message)s'
-    )
+    file_formatter = logging.Formatter('%(asctime)s [%(levelname)s] %(name)s: %(message)s')
     file_handler.setFormatter(file_formatter)
     logger.addHandler(file_handler)
 
@@ -391,10 +430,7 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
         logger.info(f'Configured activity type: {ACTIVITY_TYPE}')
         logger.info(f'Configured activity status: {ACTIVITY_STATUS}')
         activity = set_activity_status(ACTIVITY_TYPE, ACTIVITY_STATUS)
-        await bot.change_presence(
-            activity=activity,
-            status=discord.Status(BOT_PRESENCE)
-        )
+        await bot.change_presence(activity=activity, status=discord.Status(BOT_PRESENCE))
 
     @bot.event
     async def on_disconnect():
@@ -414,6 +450,9 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     async def on_shard_ready(shard_id):
         """
         Event handler for when a shard is ready.
+
+        Args:
+            shard_id: The ID of the shard.
         """
         logger.info(f'Shard {shard_id} is ready')
 
@@ -421,6 +460,9 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     async def on_message(message: discord.Message):
         """
         Event handler for when a message is received.
+
+        Args:
+            message (discord.Message): The message received.
         """
         try:
             if message.author == bot.user:

--- a/bot.py
+++ b/bot.py
@@ -12,7 +12,6 @@ from logging.handlers import RotatingFileHandler
 # Third-party imports
 import discord
 from openai import OpenAI
-from websockets.exceptions import ConnectionClosed
 
 
 class RateLimiter:
@@ -68,14 +67,9 @@ def parse_arguments() -> argparse.Namespace:
     Returns:
         argparse.Namespace: Parsed command-line arguments.
     """
-    try:
-        parser = argparse.ArgumentParser(description='GPT-based Discord bot.')
-        parser.add_argument('--conf', help='Configuration file path')
-        args = parser.parse_args()
-        return args
-    except Exception as e:
-        logger.error(f"Error parsing arguments: {e}")
-        raise
+    parser = argparse.ArgumentParser(description='GPT-based Discord bot.')
+    parser.add_argument('--conf', help='Configuration file path')
+    return parser.parse_args()
 
 
 def load_configuration(config_file: str) -> configparser.ConfigParser:
@@ -88,18 +82,14 @@ def load_configuration(config_file: str) -> configparser.ConfigParser:
     Returns:
         configparser.ConfigParser: Loaded configuration.
     """
-    try:
-        config = configparser.ConfigParser()
+    config = configparser.ConfigParser()
 
-        if os.path.exists(config_file):
-            config.read(config_file)
-        else:
-            config.read_dict({section: dict(os.environ) for section in config.sections()})
+    if os.path.exists(config_file):
+        config.read(config_file)
+    else:
+        config.read_dict({section: dict(os.environ) for section in config.sections()})
 
-        return config
-    except Exception as e:
-        logger.error(f"Error loading configuration: {e}")
-        raise
+    return config
 
 
 def set_activity_status(activity_type: str, activity_status: str) -> discord.Activity:
@@ -113,22 +103,18 @@ def set_activity_status(activity_type: str, activity_status: str) -> discord.Act
     Returns:
         discord.Activity: The activity object.
     """
-    try:
-        activity_types = {
-            'playing': discord.ActivityType.playing,
-            'streaming': discord.ActivityType.streaming,
-            'listening': discord.ActivityType.listening,
-            'watching': discord.ActivityType.watching,
-            'custom': discord.ActivityType.custom,
-            'competing': discord.ActivityType.competing
-        }
-        return discord.Activity(
-            type=activity_types.get(activity_type, discord.ActivityType.listening),
-            name=activity_status
-        )
-    except Exception as e:
-        logger.error(f"Error setting activity status: {e}")
-        raise
+    activity_types = {
+        'playing': discord.ActivityType.playing,
+        'streaming': discord.ActivityType.streaming,
+        'listening': discord.ActivityType.listening,
+        'watching': discord.ActivityType.watching,
+        'custom': discord.ActivityType.custom,
+        'competing': discord.ActivityType.competing
+    }
+    return discord.Activity(
+        type=activity_types.get(activity_type, discord.ActivityType.listening),
+        name=activity_status
+    )
 
 
 def get_conversation_summary(conversation: list[dict]) -> list[dict]:
@@ -141,19 +127,15 @@ def get_conversation_summary(conversation: list[dict]) -> list[dict]:
     Returns:
         list[dict]: The summarized conversation.
     """
-    try:
-        summary = []
-        user_messages = [msg for msg in conversation if msg["role"] == "user"]
-        assistant_responses = [msg for msg in conversation if msg["role"] == "assistant"]
+    summary = []
+    user_messages = [msg for msg in conversation if msg["role"] == "user"]
+    assistant_responses = [msg for msg in conversation if msg["role"] == "assistant"]
 
-        for user_msg, assistant_resp in zip(user_messages, assistant_responses):
-            summary.append(user_msg)
-            summary.append(assistant_resp)
+    for user_msg, assistant_resp in zip(user_messages, assistant_responses):
+        summary.append(user_msg)
+        summary.append(assistant_resp)
 
-        return summary
-    except Exception as e:
-        logger.error(f"Error getting conversation summary: {e}")
-        raise
+    return summary
 
 
 async def check_rate_limit(
@@ -179,11 +161,7 @@ async def check_rate_limit(
     if logger is None:
         logger = logging.getLogger(__name__)
 
-    try:
-        return rate_limiter.check_rate_limit(user.id, rate_limit, rate_limit_per, logger)
-    except Exception as e:
-        logger.error(f"Error checking rate limit: {e}")
-        raise
+    return rate_limiter.check_rate_limit(user.id, rate_limit, rate_limit_per, logger)
 
 
 async def process_input_message(
@@ -202,61 +180,51 @@ async def process_input_message(
     Returns:
         str: The response from the GPT model.
     """
+    logger.info("Sending prompt to the API.")
+
+    conversation = CONVERSATION_HISTORY.get(user.id, [])
+    conversation.append({"role": "user", "content": input_message})
+
+    def call_openai_api():
+        logger.debug(f"GPT_MODEL: {GPT_MODEL}")
+        logger.debug(f"SYSTEM_MESSAGE: {SYSTEM_MESSAGE}")
+        logger.debug(f"conversation_summary: {conversation_summary}")
+        logger.debug(f"input_message: {input_message}")
+
+        return client.chat.completions.create(
+            model=GPT_MODEL,
+            messages=[
+                {"role": "system", "content": SYSTEM_MESSAGE},
+                *conversation_summary,
+                {"role": "user", "content": input_message}
+            ],
+            max_tokens=OUTPUT_TOKENS,
+            temperature=0.7
+        )
+
+    response = await asyncio.to_thread(call_openai_api)
+    logger.debug(f"Full API response: {response}")
+
     try:
-        logger.info("Sending prompt to the API.")
-
-        conversation = CONVERSATION_HISTORY.get(user.id, [])
-        conversation.append({"role": "user", "content": input_message})
-
-        def call_openai_api():
-            logger.debug(f"GPT_MODEL: {GPT_MODEL}")
-            logger.debug(f"SYSTEM_MESSAGE: {SYSTEM_MESSAGE}")
-            logger.debug(f"conversation_summary: {conversation_summary}")
-            logger.debug(f"input_message: {input_message}")
-
-            return client.chat.completions.create(
-                model=GPT_MODEL,
-                messages=[
-                    {"role": "system", "content": SYSTEM_MESSAGE},
-                    *conversation_summary,
-                    {"role": "user", "content": input_message}
-                ],
-                max_tokens=OUTPUT_TOKENS,
-                temperature=0.7
-            )
-
-        response = await asyncio.to_thread(call_openai_api)
-
-        try:
-            if response.choices:
-                response_content = response.choices[0].message.content.strip()
-            else:
-                response_content = None
-        except AttributeError as e:
-            logger.error(f"Failed to get response from the API: {e}")
-            return "Sorry, an error occurred while processing the message."
-
-        if response_content:
-            logger.info("Received response from the API.")
-            logger.info(f"Sent the response: {response_content}")
-
-            conversation.append({"role": "assistant", "content": response_content})
-            CONVERSATION_HISTORY[user.id] = conversation
-
-            return response_content
+        if response.choices:
+            response_content = response.choices[0].message.content.strip()
         else:
-            logger.error("API error: No response text.")
-            return "Sorry, I didn't get that. Can you rephrase or ask again?"
+            response_content = None
+    except AttributeError as e:
+        logger.error(f"Failed to get response from the API: {e}")
+        return "Sorry, an error occurred while processing the message."
 
-    except ConnectionClosed as error:
-        logger.error(f"WebSocket connection closed: {error}")
-        logger.info("Reconnecting in 5 seconds...")
-        await asyncio.sleep(5)
-        await bot.login(DISCORD_TOKEN)
-        await bot.connect(reconnect=True)
-    except Exception as error:
-        logger.error("An error processing message: %s", error)
-        return "An error occurred while processing the message."
+    if response_content:
+        logger.info("Received response from the API.")
+        logger.info(f"Sent the response: {response_content}")
+
+        conversation.append({"role": "assistant", "content": response_content})
+        CONVERSATION_HISTORY[user.id] = conversation
+
+        return response_content
+    else:
+        logger.error("API error: No response text.")
+        return "Sorry, I didn't get that. Can you rephrase or ask again?"
 
 
 async def process_dm_message(message: discord.Message):

--- a/bot.py
+++ b/bot.py
@@ -141,10 +141,10 @@ async def process_input_message(
     conversation_summary: list[dict]
 ) -> str:
     """
-    Process an input message using OpenAI's GPT model.
+    Process an input message using the GPT model.
     """
     try:
-        logger.info("Sending prompt to OpenAI API.")
+        logger.info("Sending prompt to the API.")
 
         conversation = conversation_history.get(user.id, [])
         conversation.append({"role": "user", "content": input_message})
@@ -188,13 +188,13 @@ async def process_input_message(
                 response_content = None
         except AttributeError:
             logger.error(
-                "Failed to get response from OpenAI API: "
+                "Failed to get response from the API: "
                 "Invalid response format."
             )
             return "Sorry, an error occurred while processing the message."
 
         if response_content:
-            logger.info("Received response from OpenAI API.")
+            logger.info("Received response from the API.")
             # Debugging: Log the raw response
             # logger.info(f"Raw API response: {response}")
             logger.info(f"Sent the response: {response_content}")
@@ -206,7 +206,7 @@ async def process_input_message(
 
             return response_content
         else:
-            logger.error("OpenAI API error: No response text.")
+            logger.error("API error: No response text.")
             return "Sorry, I didn't get that. Can you rephrase or ask again?"
 
     except ConnectionClosed as error:
@@ -243,14 +243,17 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
         'Discord', 'ACTIVITY_STATUS', fallback='Humans'
         )
 
-    OPENAI_API_KEY = config.get('OpenAI', 'OPENAI_API_KEY')
-    OPENAI_TIMEOUT = config.getint('OpenAI', 'OPENAI_TIMEOUT', fallback='30')
-    GPT_MODEL = config.get(
-        'OpenAI', 'GPT_MODEL', fallback='gpt-3.5-turbo-1106'
+    # Set your Defaults
+    API_KEY = config.get('Default', 'API_KEY')
+    API_URL = config.get(
+        'Default', 'API_URL', fallback='https://api.openai.com/v1/'
     )
-    GPT_TOKENS = config.getint('OpenAI', 'GPT_TOKENS', fallback=4096)
+    GPT_MODEL = config.get(
+        'Default', 'GPT_MODEL', fallback='gpt-3.5-turbo-1106'
+    )
+    GPT_TOKENS = config.getint('Default', 'GPT_TOKENS', fallback=4096)
     SYSTEM_MESSAGE = config.get(
-        'OpenAI', 'SYSTEM_MESSAGE', fallback='You are a helpful assistant.'
+        'Default', 'SYSTEM_MESSAGE', fallback='You are a helpful assistant.'
     )
 
     RATE_LIMIT = config.getint('Limits', 'RATE_LIMIT', fallback=10)
@@ -299,9 +302,10 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     # Create the bot instance
     bot = discord.Client(intents=intents)
 
-    # Create the OpenAI client instance
+    # Create the API client instance
     client = OpenAI(
-        api_key=OPENAI_API_KEY
+        api_key=API_KEY,
+        api_base=API_URL
     )
 
     @bot.event

--- a/bot.py
+++ b/bot.py
@@ -14,6 +14,29 @@ import discord
 from openai import OpenAI
 from websockets.exceptions import ConnectionClosed
 
+class RateLimiter:
+    def __init__(self):
+        self.last_command_timestamps = {}
+        self.last_command_count = {}
+
+    def check_rate_limit(self, user_id: int, rate_limit: int, rate_limit_per: int, logger: logging.Logger) -> bool:
+        current_time = time.time()
+        last_command_timestamp = self.last_command_timestamps.get(user_id, 0)
+        last_command_count_user = self.last_command_count.get(user_id, 0)
+
+        if current_time - last_command_timestamp > rate_limit_per:
+            self.last_command_timestamps[user_id] = current_time
+            self.last_command_count[user_id] = 1
+            logger.info(f"Rate limit passed for user: {user_id}")
+            return True
+
+        if last_command_count_user < rate_limit:
+            self.last_command_count[user_id] += 1
+            logger.info(f"Rate limit passed for user: {user_id}")
+            return True
+
+        logger.info(f"Rate limit exceeded for user: {user_id}")
+        return False
 
 # Define the function to parse command-line arguments
 def parse_arguments() -> argparse.Namespace:
@@ -25,7 +48,6 @@ def parse_arguments() -> argparse.Namespace:
     except Exception as e:
         logger.error(f"Error parsing arguments: {e}")
         raise
-
 
 # Define the function to load the configuration
 def load_configuration(config_file: str) -> configparser.ConfigParser:
@@ -45,7 +67,6 @@ def load_configuration(config_file: str) -> configparser.ConfigParser:
     except Exception as e:
         logger.error(f"Error loading configuration: {e}")
         raise
-
 
 def set_activity_status(
     activity_type: str,
@@ -73,7 +94,6 @@ def set_activity_status(
         logger.error(f"Error setting activity status: {e}")
         raise
 
-
 # Define the function to get the conversation summary
 def get_conversation_summary(conversation: list[dict]) -> list[dict]:
     """
@@ -100,9 +120,11 @@ def get_conversation_summary(conversation: list[dict]) -> list[dict]:
         logger.error(f"Error getting conversation summary: {e}")
         raise
 
-
 async def check_rate_limit(
     user: discord.User,
+    rate_limiter: RateLimiter,
+    rate_limit: int,
+    rate_limit_per: int,
     logger: logging.Logger = None
 ) -> bool:
     """
@@ -112,28 +134,10 @@ async def check_rate_limit(
         logger = logging.getLogger(__name__)
 
     try:
-        current_time = time.time()
-        last_command_timestamp = last_command_timestamps.get(user.id, 0)
-        last_command_count_user = last_command_count.get(user.id, 0)
-
-        if current_time - last_command_timestamp > RATE_LIMIT_PER:
-            last_command_timestamps[user.id] = current_time
-            last_command_count[user.id] = 1
-            logger.info(f"Rate limit passed for user: {user}")
-            return True
-
-        if last_command_count_user < RATE_LIMIT:
-            last_command_count[user.id] += 1
-            logger.info(f"Rate limit passed for user: {user}")
-            return True
-
-        logger.info(f"Rate limit exceeded for user: {user}")
-        return False
-
+        return rate_limiter.check_rate_limit(user.id, rate_limit, rate_limit_per, logger)
     except Exception as e:
         logger.error(f"Error checking rate limit: {e}")
         raise
-
 
 async def process_input_message(
     input_message: str,
@@ -145,6 +149,8 @@ async def process_input_message(
     """
     try:
         logger.info("Sending prompt to the API.")
+
+        INPUT_PROMPT = {"role": "user", "content": input_message}
 
         conversation = conversation_history.get(user.id, [])
         conversation.append({"role": "user", "content": input_message})
@@ -219,7 +225,6 @@ async def process_input_message(
         logger.error("An error processing message: %s", error)
         return "An error occurred while processing the message."
 
-
 # Executes the argparse code only when the file is run directly
 if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     # Parse command-line arguments
@@ -292,13 +297,12 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     intents.typing = False
     intents.presences = False
 
-    # Create a dictionary to store the last command timestamp for each user
-    last_command_timestamps = {}
-    last_command_count = {}
-
     # Create a dictionary to store conversation history for each user
     conversation_history = {}
 
+    # Setup the Bot's Personality
+    SYSTEM_PROMPT = {"role": "system", "content": SYSTEM_MESSAGE}
+    
     # Create the bot instance
     bot = discord.Client(intents=intents)
 
@@ -307,6 +311,9 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
         api_key=API_KEY,
         api_base=API_URL
     )
+
+    # Initialize rate limiter
+    rate_limiter = RateLimiter()
 
     @bot.event
     async def on_ready():
@@ -372,7 +379,7 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
             f'Received DM from {message.author}: {message.content}'
         )
 
-        if not await check_rate_limit(message.author):
+        if not await check_rate_limit(message.author, rate_limiter, RATE_LIMIT, RATE_LIMIT_PER):
             await message.channel.send(
                 f"{message.author.mention} Exceeded the Rate Limit! Please slow down!"
             )
@@ -399,7 +406,7 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
             )
         )
 
-        if not await check_rate_limit(message.author):
+        if not await check_rate_limit(message.author, rate_limiter, RATE_LIMIT, RATE_LIMIT_PER):
             await message.channel.send(
                 f"{message.author.mention} Exceeded the Rate Limit! Please slow down!"
             )

--- a/bot.py
+++ b/bot.py
@@ -164,24 +164,29 @@ async def process_input_message(
     try:
         logger.info("Sending prompt to the API.")
 
-        INPUT_PROMPT = {"role": "user", "content": input_message}
-
         conversation = CONVERSATION_HISTORY.get(user.id, [])
-        conversation.append(INPUT_PROMPT)
+        conversation.append({"role": "user", "content": input_message})
 
         def call_openai_api():
+            logger.debug(f"GPT_MODEL: {GPT_MODEL}")
+            logger.debug(f"SYSTEM_MESSAGE: {SYSTEM_MESSAGE}")
+            logger.debug(f"conversation_summary: {conversation_summary}")
+            logger.debug(f"input_message: {input_message}")
+
             return client.chat.completions.create(
                 model=GPT_MODEL,
                 messages=[
-                    SYSTEM_PROMPT,
+                    {"role": "system", "content": SYSTEM_MESSAGE},
                     *conversation_summary,
-                    INPUT_PROMPT
+                    {"role": "user", "content": input_message}
                 ],
                 max_tokens=OUTPUT_TOKENS,
                 temperature=0.7
             )
 
         response = await asyncio.to_thread(call_openai_api)
+        # Debugging: Log the raw response
+        # logger.info(f"Raw API response: {response}")
 
         try:
             # Extracting the response content from the new API response format
@@ -189,15 +194,14 @@ async def process_input_message(
                 response_content = response.choices[0].message.content.strip()
             else:
                 response_content = None
-        except AttributeError:
-            logger.error(
-                "Failed to get response from the API: "
-                "Invalid response format."
-            )
+        except AttributeError as e:
+            logger.error(f"Failed to get response from the API: {e}")
             return "Sorry, an error occurred while processing the message."
 
         if response_content:
             logger.info("Received response from the API.")
+            # Debugging: Log the raw response
+            # logger.info(f"Raw API response: {response}")
             logger.info(f"Sent the response: {response_content}")
 
             conversation.append(
@@ -219,6 +223,78 @@ async def process_input_message(
     except Exception as error:
         logger.error("An error processing message: %s", error)
         return "An error occurred while processing the message."
+
+
+async def process_dm_message(message: discord.Message):
+    """
+    Process a direct message.
+    """
+    logger.info(f'Received DM from {message.author}: {message.content}')
+
+    if not await check_rate_limit(message.author, rate_limiter, RATE_LIMIT, RATE_LIMIT_PER):
+        await message.channel.send(
+            f"{message.author.mention} Exceeded the Rate Limit! Please slow down!"
+        )
+        logger.warning(f"Rate Limit Exceeded by DM from {message.author}")
+        return
+
+    conversation_summary = get_conversation_summary(
+        CONVERSATION_HISTORY.get(message.author.id, [])
+    )
+    response = await process_input_message(
+        message.content, message.author, conversation_summary
+    )
+    await send_split_message(message.channel, response)
+
+
+async def process_channel_message(message: discord.Message):
+    """
+    Process a message in a channel.
+    """
+    logger.info(
+        'Received message in {} from {}: {}'.format(
+            str(message.channel),
+            str(message.author),
+            re.sub(r'<@\d+>', '', message.content)
+        )
+    )
+
+    if not await check_rate_limit(message.author, rate_limiter, RATE_LIMIT, RATE_LIMIT_PER):
+        await message.channel.send(
+            f"{message.author.mention} Exceeded the Rate Limit! Please slow down!"
+        )
+        logger.warning(f"Rate Limit Exceeded in {message.channel} by {message.author}")
+        return
+
+    conversation_summary = get_conversation_summary(
+        CONVERSATION_HISTORY.get(message.author.id, [])
+    )
+    response = await process_input_message(
+        message.content, message.author, conversation_summary
+    )
+    await send_split_message(message.channel, response)
+
+
+async def send_split_message(channel: discord.abc.Messageable, message: str):
+    """
+    Send a message to a channel. If the message is longer than 2000 characters,
+    it is split into multiple messages at the nearest newline character around
+    the middle of the message.
+    """
+    if len(message) <= 2000:
+        await channel.send(message)
+    else:
+        # Find the nearest newline character around the middle of the message
+        middle_index = len(message) // 2
+        split_index = message.rfind('\n', 0, middle_index)
+        if split_index == -1:  # No newline character found
+            split_index = middle_index  # Split at the middle of the message
+        # Split the message into two parts
+        message_part1 = message[:split_index]
+        message_part2 = message[split_index:]
+        # Send the two parts as separate messages
+        await channel.send(message_part1)
+        await channel.send(message_part2)
 
 # Executes the argparse code only when the file is run directly
 if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
@@ -296,17 +372,11 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
     # Create a dictionary to store conversation history for each user
     CONVERSATION_HISTORY = {}
 
-    # Setup the Bot's Personality
-    SYSTEM_PROMPT = {"role": "system", "content": SYSTEM_MESSAGE}
-
     # Create the bot instance
     bot = discord.Client(intents=intents)
 
-    # Create the API client instance
-    client = OpenAI(
-        api_key=API_KEY,
-        api_base=API_URL
-    )
+    # Set the API key and base URL
+    client = OpenAI(api_key=API_KEY, base_url=API_URL)
 
     # Initialize rate limiter
     rate_limiter = RateLimiter()
@@ -348,7 +418,7 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
         logger.info(f'Shard {shard_id} is ready')
 
     @bot.event
-    async def on_message(message):
+    async def on_message(message: discord.Message):
         """
         Event handler for when a message is received.
         """
@@ -366,77 +436,6 @@ if __name__ == "__main__":  # noqa: C901 (ignore complexity in main function)
                 await process_channel_message(message)
         except Exception as e:
             logger.error(f"An error occurred in on_message: {e}")
-
-    async def process_dm_message(message):
-        """
-        Process a direct message.
-        """
-        logger.info(
-            f'Received DM from {message.author}: {message.content}'
-        )
-
-        if not await check_rate_limit(message.author, rate_limiter, RATE_LIMIT, RATE_LIMIT_PER):
-            await message.channel.send(
-                f"{message.author.mention} Exceeded the Rate Limit! Please slow down!"
-            )
-            logger.warning(f"Rate Limit Exceed by DM from {message.author}")
-            return
-
-        conversation_summary = get_conversation_summary(
-            CONVERSATION_HISTORY.get(message.author.id, [])
-        )
-        response = await process_input_message(
-            message.content, message.author, conversation_summary
-        )
-        await send_split_message(message.channel, response)
-
-    async def process_channel_message(message):
-        """
-        Process a message in a channel.
-        """
-        logger.info(
-            'Received message in {} from {}: {}'.format(
-                str(message.channel),
-                str(message.author),
-                re.sub(r'<@\d+>', '', message.content)
-            )
-        )
-
-        if not await check_rate_limit(message.author, rate_limiter, RATE_LIMIT, RATE_LIMIT_PER):
-            await message.channel.send(
-                f"{message.author.mention} Exceeded the Rate Limit! Please slow down!"
-            )
-            logger.warning(f"Rate Limit Exceeded in {message.channel} by {message.author}")
-            return
-
-        conversation_summary = get_conversation_summary(
-            CONVERSATION_HISTORY.get(message.author.id, [])
-        )
-        response = await process_input_message(
-            message.content, message.author, conversation_summary
-        )
-        await send_split_message(message.channel, response)
-
-    async def send_split_message(channel, message):
-        """
-        Send a message to a channel. If the message is longer than 2000 characters,
-        it is split into multiple messages at the nearest newline character around
-        the middle of the message.
-        """
-        if len(message) <= 2000:
-            await channel.send(message)
-        else:
-            # Find the nearest newline character around the middle of the message
-            middle_index = len(message) // 2
-            split_index = message.rfind('\n', 0, middle_index)
-            if split_index == -1:  # No newline character found
-                split_index = middle_index  # Split at the middle of the message
-            # Split the message into two parts
-            message_part1 = message[:split_index]
-            message_part2 = message[split_index:]
-            # Send the two parts as separate messages
-            await channel.send(message_part1)
-            await channel.send(message_part2)
 
     # Run the bot
     bot.run(DISCORD_TOKEN)

--- a/config.ini.example
+++ b/config.ini.example
@@ -7,11 +7,13 @@ BOT_PRESENCE = online
 ACTIVITY_TYPE=listening
 ACTIVITY_STATUS=Humans
 
-[OpenAI]
-OPENAI_API_KEY = <your_openai_api_key>
-OPENAI_TIMEOUT=30
-GPT_MODEL=gpt-3.5-turbo
-GPT_TOKENS=3072
+[Default]
+API_URL=https://api.openai.com/v1/
+API_KEY = <your_api_key>
+GPT_MODEL=gpt-4o
+INPUT_TOKENS=124000
+OUTPUT_TOKENS=4000
+CONTEXT_WINDOW=128000
 SYSTEM_MESSAGE = You are a helpful AI assistant.
 
 [Limits]

--- a/config.ini.example
+++ b/config.ini.example
@@ -10,9 +10,9 @@ ACTIVITY_STATUS=Humans
 [Default]
 API_URL=https://api.openai.com/v1/
 API_KEY = <your_api_key>
-GPT_MODEL=gpt-4o
-INPUT_TOKENS=124000
-OUTPUT_TOKENS=4000
+GPT_MODEL=gpt-4o-mini
+INPUT_TOKENS=120000
+OUTPUT_TOKENS=8000
 CONTEXT_WINDOW=128000
 SYSTEM_MESSAGE = You are a helpful AI assistant.
 

--- a/pytest.ini
+++ b/pytest.ini
@@ -1,3 +1,5 @@
 [pytest]
 addopts =
     -ra -q
+
+asyncio_default_fixture_loop_scope = function

--- a/tests/test_check_rate_limit.py
+++ b/tests/test_check_rate_limit.py
@@ -1,12 +1,8 @@
-from contextlib import contextmanager
 import time
-from unittest.mock import AsyncMock
-import pytest
 import logging
-
-import bot
-from bot import check_rate_limit
-
+import pytest
+from unittest.mock import AsyncMock
+from bot import RateLimiter
 
 # Define a placeholder logger
 logger = logging.getLogger('pytest_logger')
@@ -18,56 +14,27 @@ logger.addHandler(handler)
 RATE_LIMIT = 10
 RATE_LIMIT_PER = 60
 
-last_command_count = {}
-last_command_timestamps = {}
+def run_test(user, rate_limiter):
+    # Simulate first command
+    result = rate_limiter.check_rate_limit(user.id, RATE_LIMIT, RATE_LIMIT_PER, logger)
+    assert result is True
+    assert rate_limiter.last_command_count.get(user.id, 0) == 1
 
+    # Simulate reaching rate limit
+    rate_limiter.last_command_count[user.id] = RATE_LIMIT
+    result = rate_limiter.check_rate_limit(user.id, RATE_LIMIT, RATE_LIMIT_PER, logger)
+    assert result is False
+    assert rate_limiter.last_command_count.get(user.id, 0) == RATE_LIMIT
 
-@contextmanager
-def patch_variables(module, variable_name, value):
-    """Temporarily patches a variable in a module with a given value."""
-    original_value = getattr(module, variable_name, None)
-    setattr(module, variable_name, value)
-    try:
-        yield
-    finally:
-        setattr(module, variable_name, original_value)
-
-
-async def run_test(user):
-    with patch_variables(
-        bot,
-        'last_command_timestamps',
-        {user.id: time.time() - 60}
-    ), patch_variables(
-        bot,
-        'last_command_count',
-        {user.id: 0}
-    ), patch_variables(
-        bot,
-        'RATE_LIMIT_PER',
-        RATE_LIMIT_PER
-    ), patch_variables(
-        bot,
-        'RATE_LIMIT',
-        RATE_LIMIT
-    ):
-        result = await check_rate_limit(user, logger)
-        assert result is True
-        assert bot.last_command_count.get(user.id, 0) == 1
-
-        bot.last_command_count[user.id] = RATE_LIMIT
-        result = await check_rate_limit(user, logger)
-        assert result is False
-        assert bot.last_command_count.get(user.id, 0) == RATE_LIMIT
-
-        bot.last_command_timestamps[user.id] = time.time() - RATE_LIMIT_PER - 1
-        result = await check_rate_limit(user, logger)
-        assert result is True
-        assert bot.last_command_count.get(user.id, 0) == 1
-
+    # Simulate time passing to reset rate limit
+    rate_limiter.last_command_timestamps[user.id] = time.time() - RATE_LIMIT_PER - 1
+    result = rate_limiter.check_rate_limit(user.id, RATE_LIMIT, RATE_LIMIT_PER, logger)
+    assert result is True
+    assert rate_limiter.last_command_count.get(user.id, 0) == 1
 
 @pytest.mark.asyncio
 async def test_check_rate_limit():
     user = AsyncMock()
     user.id = 123
-    await run_test(user)
+    rate_limiter = RateLimiter()
+    run_test(user, rate_limiter)

--- a/tests/test_check_rate_limit.py
+++ b/tests/test_check_rate_limit.py
@@ -14,6 +14,7 @@ logger.addHandler(handler)
 RATE_LIMIT = 10
 RATE_LIMIT_PER = 60
 
+
 def run_test(user, rate_limiter):
     # Simulate first command
     result = rate_limiter.check_rate_limit(user.id, RATE_LIMIT, RATE_LIMIT_PER, logger)
@@ -31,6 +32,7 @@ def run_test(user, rate_limiter):
     result = rate_limiter.check_rate_limit(user.id, RATE_LIMIT, RATE_LIMIT_PER, logger)
     assert result is True
     assert rate_limiter.last_command_count.get(user.id, 0) == 1
+
 
 @pytest.mark.asyncio
 async def test_check_rate_limit():

